File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/test/HardwareTest.java
package org.firstinspires.ftc.teamcode.opmode.teleop.test;

import com.qualcomm.robotcore.eventloop.opmode.TeleOp;
import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.hardware.Servo;

import org.firstinspires.ftc.teamcode.modules.UI.BoolEx;
import org.firstinspires.ftc.teamcode.modules.UI.FloatEx;
import org.firstinspires.ftc.teamcode.modules.UI.IntEx;
import org.firstinspires.ftc.teamcode.modules.UI.UI;
import org.firstinspires.ftc.teamcode.opmode.teleop.BaseTeleOp;

@TeleOp
public class HardwareTest extends BaseTeleOp {
    UI ui;
    int mode = 0;
    DcMotor motor;
    Servo servo;
    IntEx deviceNumber;
    FloatEx power;
    BoolEx showAll;
    public void init() {
        super.init();
        showAll = new BoolEx();
        ui = new UI(telemetry, gamepadEx1);
        power = new FloatEx();
        deviceNumber = new IntEx();
    }
    public void loop() {
        ui.update();
        if (mode != 0) {
            if (ui.button("back")) {
                mode = 0;
                ui.resetSelected();
            }
            ui.checkbox("show all", showAll);
            switch (mode) {
                case 1:
                    if (showAll.value) {
                        drawAllDcMotors();
                    } else {
                        drawDcMotor();
                    }
                    break;
                case 2:
                    if(showAll.value) {
                        drawAllServos();
                    }else {
                        drawServo();
                    }
                    break;
            }
        } else {
            if (ui.button("DcMotor")) {
                mode = 1;
                motor = hardwareMap.dcMotor.iterator().next();
                ui.resetSelected();
            }
            if (ui.button("Servo")) {
                mode = 2;
                servo = hardwareMap.servo.iterator().next();
                ui.resetSelected();
            }
        }
        telemetry.update();
    }
    void drawDcMotor(){
        if(ui.intInput("motor number", deviceNumber, 1)){
            int i = 0;
            for (DcMotor motor1 : hardwareMap.dcMotor) {
                if(i == deviceNumber.value){
                    motor = motor1;
                    break;
                }
                i++;
            }
        }
        ui.label("port: " + motor.getPortNumber());
        ui.label("connection info", motor.getConnectionInfo());
        ui.label("current position", motor.getCurrentPosition());
        if(ui.floatInput("power", power, 0.05f)){
            motor.setPower(power.value);
        }
        if(ui.button("stop")){
            motor.setPower(0);
            power.value = 0.0f;
        }
    }
    void drawServo(){
        if(ui.intInput("servo number", deviceNumber, 1)){
            int i = 0;
            for (Servo servo1: hardwareMap.servo) {
                if(i == deviceNumber.value){
                    servo = servo1;
                    break;
                }
                i++;
            }
        }
        ui.label("port: " + servo.getPortNumber());
        ui.label("connection info", servo.getConnectionInfo());
        ui.label("current position", servo.getPosition());
        if(ui.floatInput("targetPosition", power, 0.05f)){
            servo.setPosition(power.value);
        }
        if(ui.button("zero")){
            servo.setPosition(0);
            power.value = 0.0f;
        }
    }
    void drawAllDcMotors(){
        for(DcMotor motor1 : hardwareMap.dcMotor){
            ui.label("port: " + motor1.getPortNumber());
            ui.label("connection info", motor1.getConnectionInfo());
            ui.label("current power", motor1.getPower());
            ui.label("current position", motor1.getCurrentPosition());
        }
    }
    void drawAllServos(){
        for(Servo servo1 : hardwareMap.servo){
            ui.label("port: " + servo1.getPortNumber());
            ui.label("connection info", servo1.getConnectionInfo());
            ui.label("current position", servo1.getPosition());
        }
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/test/HardwareTest.kt
package org.firstinspires.ftc.teamcode.opmode.teleop.test

import com.qualcomm.robotcore.eventloop.opmode.OpMode
import com.qualcomm.robotcore.eventloop.opmode.TeleOp
import com.qualcomm.robotcore.hardware.DcMotorEx
import com.qualcomm.robotcore.hardware.Servo
import org.firstinspires.ftc.robotcore.external.navigation.CurrentUnit
import org.firstinspires.ftc.teamcode.modules.Ui.FloatPtr
import org.firstinspires.ftc.teamcode.modules.Ui.BoolPtr
import org.firstinspires.ftc.teamcode.modules.Ui.IntPtr
import org.firstinspires.ftc.teamcode.modules.Ui.UI

@TeleOp
class HardwareTest : OpMode {
    private enum class Mode{
        Motor, Servo, None
    }
    private var ui: UI = UI();
    private var mode: Mode = Mode.None;

    private var index: IntPtr = IntPtr();

    private var motor: DcMotorEx? = null;
    private var motorList: MutableList<DcMotorEx> = ArrayList();
    private val motorPower: FloatPtr = FloatPtr();
    private val motorTargetPosition: IntPtr = IntPtr();
    private val motorUpdateTargetPosition: BoolPtr = BoolPtr();


    private var servo: Servo? = null;
    private var servoList: MutableList<Servo> = ArrayList();
    private val servoPosition: FloatPtr = FloatPtr();

    constructor(){}
    override fun init(){
        ui.init(telemetry, gamepad1)
        hardwareMap.forEach() {
            if(it is DcMotorEx){
                motorList.add(it);
            }
            if(it is Servo){
                servoList.add(it);
            }
        }
    }
    override fun start(){}
    override fun loop(){
        when(mode) {
            Mode.None->{
                if (ui.button("motor")) {
                    mode = Mode.Motor;
                    index.value = 0;
                    motor = motorList[index.value];
                }
                if (ui.button("servo")) {
                    mode = Mode.Servo;
                    index.value = 0;
                    servo = servoList[index.value];
                }
            }
            Mode.Motor->{
                if(ui.button("back")) {
                    mode = Mode.None;
                }
                if(ui.intInput("index", index, 1)){
                    if(index.value >= motorList.size) {
                        index.value = motorList.size - 1;
                    }
                    if(index.value <= 0) {
                        index.value = 0;
                    }
                    motor = motorList[index.value];
                }
                ui.label("name" , motor?.connectionInfo);
                ui.label("port" , motor?.portNumber);
                ui.label("currentPosition", motor?.currentPosition);
                ui.label("velocity", motor?.velocity);
                ui.label("current", motor?.getCurrent(CurrentUnit.MILLIAMPS));
                if(ui.floatInput("power", motorPower, 0.1f)) {
                    motor?.power = motorPower.value.toDouble();
                }
                if(ui.button("stop")){
                    motor?.power = 0.0;
                    motorPower.value = 0.0f;
                }
                if(ui.checkbox("updateTargetPosition", motorUpdateTargetPosition)){
                  motor?.targetPosition = motorTargetPosition.value;
                }
                if(ui.intInput("targetPosition", motorTargetPosition, 50)){
                  if(motorUpdateTargetPosition.value){
                    motor?.targetPosition = motorTargetPosition.value;
                  }
                } 
            }
            Mode.Servo->{
                if(ui.button("back")) {
                    mode = Mode.None;
                }
                if(ui.intInput("index", index, 1)){
                    if(index.value >= servoList.size) {
                        index.value = servoList.size - 1;
                    }
                    if(index.value <= 0) {
                        index.value = 0;
                    }
                    servo = servoList[index.value];
                }
                ui.label("name" , servo?.connectionInfo);
                ui.label("port" , servo?.portNumber);
                if(ui.floatInput("position", servoPosition, 0.05f)) {
                    servo?.position = motorPower.value.toDouble();
                }
                if(ui.button("reset position")){
                    servo?.position = 0.0;
                    servoPosition.value = 0.0f;
                }
            }
        }
        ui.update();
    }
  }
